src/versions.jl

```diff
@@ -0,0 +1,299 @@
+VersionSpec(s::AbstractString) = VersionSpec(VersionRange(s))
+VersionSpec(v::AbstractVector) = VersionSpec(map(VersionRange, v))
+
+# Hot code
+function Base.in(v::VersionNumber, s::VersionSpec)
+    for r in s.ranges
+        v in r && return true
+    end
+    return false
+end
+
+Base.copy(vs::VersionSpec) = VersionSpec(vs)
+
+const empty_versionspec = VersionSpec(VersionRange[])
+# Windows console doesn't like Unicode
+const _empty_symbol = @static Sys.iswindows() ? "empty" : "∅"
+
+Base.isempty(s::VersionSpec) = all(isempty, s.ranges)
+@assert isempty(empty_versionspec)
+# Hot code, measure performance before changing
+function Base.intersect(A::VersionSpec, B::VersionSpec)
+    (isempty(A) || isempty(B)) && return copy(empty_versionspec)
+    ranges = Vector{VersionRange}(undef, length(A.ranges) * length(B.ranges))
+    i = 1
+    @inbounds for a in A.ranges, b in B.ranges
+        ranges[i] = intersect(a, b)
+        i += 1
+    end
+    VersionSpec(ranges)
+end
+Base.intersect(a::VersionNumber, B::VersionSpec) = a in B ? VersionSpec(a) : empty_versionspec
+Base.intersect(A::VersionSpec, b::VersionNumber) = intersect(b, A)
+
+Base.union(A::VersionSpec, B::VersionSpec) = union!(copy(A), B)
+function Base.union!(A::VersionSpec, B::VersionSpec)
+    A == B && return A
+    append!(A.ranges, B.ranges)
+    union!(A.ranges)
+    return A
+end
+
+Base.:(==)(A::VersionSpec, B::VersionSpec) = A.ranges == B.ranges
+Base.hash(s::VersionSpec, h::UInt) = hash(s.ranges, h + (0x2fd2ca6efa023f44 % UInt))
+Base.deepcopy_internal(vs::VersionSpec, ::IdDict) = copy(vs)
+
+function Base.print(io::IO, s::VersionSpec)
+    isempty(s) && return print(io, _empty_symbol)
+    length(s.ranges) == 1 && return print(io, s.ranges[1])
+    print(io, '[')
+    for i = 1:length(s.ranges)
+        1 < i && print(io, ", ")
+        print(io, s.ranges[i])
+    end
+    print(io, ']')
+end
+Base.show(io::IO, s::VersionSpec) = print(io, "VersionSpec(\"", s, "\")")
+
+
+###################
+# Semver notation #
+###################
+
+const ver_reg = r"^([~^]?)?[v]?([0-9]+?)(?:\.([0-9]+?))?(?:\.([0-9]+?))?$"
+
+function semver_spec(s::String)
+    ranges = VersionRange[]
+    for ver in split(s, ',')
+        push!(ranges, semver_interval(ver))
+    end
+    return VersionSpec(ranges)
+end
+
+function semver_interval(s::AbstractString)
+    if !contains(s, ver_reg)
+        error("invalid version specifier")
+    end
+    m = match(ver_reg, s)
+    @assert length(m.captures) == 4
+    n_significant = count(x -> x !== nothing, m.captures) - 1
+    typ, _major, _minor, _patch = m.captures
+    major =                           parse(Int, _major)
+    minor = (n_significant < 2) ? 0 : parse(Int, _minor)
+    patch = (n_significant < 3) ? 0 : parse(Int, _patch)
+    if n_significant == 3 && major == 0 && minor == 0 && patch == 0
+        error("invalid version 0.0.0")
+    end
+    vertyp = (typ == "" || typ == "^") ? :caret : :tilde
```

Comment:
Might want a comment here indicating that this makes `^` the default type specifier similar to Cargo but dissimilar from NPM.